transfers/stager.go

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2023 The KBase Project and its Contributors
+// Copyright (c) 2023 Cohere Consulting, LLC
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of
+// this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to
+// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+// of the Software, and to permit persons to whom the Software is furnished to do
+// so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package transfers
+
+import (
+	"encoding/gob"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/kbase/dts/databases"
+)
+
+//--------
+// Stager
+//--------
+
+// The stager coordinates the staging of files at a source database in preparation for transfer.
+
+// stager global state
+var stager stagerState
+
+type stagerState struct {
+	Channels  stagerChannels
+	Databases map[string]databases.Database
+}
+
+type stagerChannels struct {
+	RequestStaging      chan uuid.UUID
+	RequestCancellation chan uuid.UUID
+	Error               chan error
+	SaveAndStop         chan *gob.Encoder
+}
+
+func newStagerChannels() stagerChannels {
+	return stagerChannels{
+		RequestStaging:      make(chan uuid.UUID, 31),
+		RequestCancellation: make(chan uuid.UUID, 31),
+		Error:               make(chan error, 31),
+		SaveAndStop:         make(chan *gob.Encoder),
+	}
+}
+
+func (channels *stagerChannels) close() {
+	close(channels.RequestStaging)
+	close(channels.RequestCancellation)
+	close(channels.Error)
+	close(channels.SaveAndStop)
+}
+
+// starts the stager
+func (s *stagerState) Start() error {
+	slog.Debug("stager.Start")
+	s.Channels = newStagerChannels()
+	go s.process(nil)
+	return <-s.Channels.Error
+}
+
+// loads the stager from saved data
+func (s *stagerState) Load(decoder *gob.Decoder) error {
+	slog.Debug("stager.Load")
+	s.Channels = newStagerChannels()
+	go s.process(decoder)
+	return <-s.Channels.Error
+}
+
+// stops the stager goroutine
+func (s *stagerState) SaveAndStop(encoder *gob.Encoder) error {
+	slog.Debug("stager.Stop")
+	s.Channels.SaveAndStop <- encoder
+	err := <-s.Channels.Error
+	s.Channels.close()
+	return err
+}
+
+// requests that files be staged for the transfer with the given ID
```

Comment:
```suggestion
// requests that files be staged and moved for the transfer with the given ID
```
It took me a little bit to figure out how the move is triggered when files need to be staged first. I'm not sure if this is the best place to put it, but a comment that indicates the periodic checks for staging status will trigger a move once a stage is complete might be helpful

Reply:
change 'will be' to 'is' and you've got yourself a deal!